src/Week6/Fibonacci.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE FlexibleInstances #-}
+module Week6.Fibonacci where
+
+
+---------- ex 1 ----------
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n-1) + fib (n-2)
+
+fibs1 :: [Integer]
+fibs1 = map fib [0..]
+
+---------- ex 2 ----------
+fibs2 :: [Integer]
+fibs2 = [0, 1] ++ fibh 0 1
+    where
+        fibh :: Integer -> Integer -> [Integer]
+        fibh x y = (x+y) : (fibh y (x+y))
+
+---------- ex 3 ----------
+data Stream a =
+    Stream a (Stream a)
+
+streamToList :: Stream a -> [a]
+streamToList (Stream x sx) = x : streamToList sx
+
+instance Show a => Show (Stream a) where
+    show = show . take 40 . streamToList
+
+---------- ex 4 ----------
+streamRepeat :: a -> Stream a
+streamRepeat x = Stream x (streamRepeat x)
```

Comment:
I mentioned previosly on one of my lectures about lazy evaluation that the following implementation for function like repeat is inefficient.

```haskell
repeat x = x : repeat x
```

Instead, it's more efficient to write:

```haskell
repeat x = let xs = x : xs in xs
```

You could to implement `streamRepeat` in similar way 🙂